packages/xstate-fsm/test/fsm.test.ts

```diff
@@ -358,6 +365,30 @@ describe('interpreter', () => {
     expect(executed).toBe(true);
   });
 
+  it('should execute exit action', () => {
+    let executed = false;
+    let service;
+    const machine = createMachine({
+      initial: 'foo',
+      states: {
+        foo: {
+          exit: () => {
+            executed = service.state.value === 'foo';
+          },
```

Comment:
It's not exactly something we are inclined to support. You shouldn't be checking `service.state` in your actions like this. A machine is like a class whereas the service is like an instance of that class - checking the instance's state through an external reference doesn't make a lot of sense.

Reply:
To add a little more context. [This issue](https://github.com/statelyai/xstate/issues/4311) was raised as we noticed this behaviour in the full package in addition to the `@xstate/fsm` package. We use the `@xstate/fsm` package in our projects.

The response on the issue is that it has been fixed in v5 but we couldn't see a fix for the fsm package. Hence the PR proposing a fix for it.

Reply:
I understand but how those things are executed internally is kinda of an implementation detail really. Conceptually the exit actions only have to be executed before transition actions - that's an observable effect that we care about.

Reply:
Isn't the response to [the issue](https://github.com/statelyai/xstate/issues/4311) raised an acknowledgement that exit actions should run **before** state changes?

Reply:
The issue that you mentioned wasn't exactly "fixed" in v5. The fact that `.getSnapshot()` within an exit action still refers to the previous state is also just an implementation detail that semi-leaks to such a call. After all, we need to update the internal data structures that this reaches into *at some point* and naturally it might be observable one way or another. However, relying on those details is not an idiomatic way of using XState.

To achieve semantics parity between `xstate@beta` and `@xstate/fsm` here you'd have to execute actions within the `machine.transition` call - instead of executing them **after** that call. That would break other scenarios though, like this one:
```ts
let executed = false;
let service;
const machine = createMachine({
  initial: "foo",
  states: {
    foo: {
      on: {
        BAR: "bar",
      },
    },
    bar: {
      entry: () => {
        executed = service.state.value === "bar";
      },
    },
  },
});

service = interpret(machine);
service.start();
service.send("BAR");
expect(executed).toBe(true);
```

This assertion passes today but it would fail with the change that I mentioned (but it would match the equivalent `xstate@beta` code).

For those reasons, I don't think it's something that we'd like to "fix". it's just not something that you should rely on in the first place.

Reply:
Ok, I understand. Thanks for clarifying @Andarist